Search Scroll witch keep alive prevent shards from closing (during Index closing)

Say my scroll keep alive time is set to 1 minute.
1. I'm executing a search with that scroll settings
2. Immediately after receiving my search result I'm closing and reopening the index
3. For exactly one minute I get the following error message and my cluster is red
4. after 1 minute my cluster is green.

 ```
[2015-06-22 14:48:13,846][INFO ][cluster.metadata         ] [Grim Hunter] closing indices [[de_v4]]
[2015-06-22 14:48:22,263][INFO ][cluster.metadata         ] [Grim Hunter] opening indices [[de_v4]]
[2015-06-22 14:48:27,308][WARN ][indices.cluster          ] [Grim Hunter] [[de_v4][2]] marking and sending shard failed due to [failed to create shard]
org.elasticsearch.index.shard.IndexShardCreationException: [de_v4][2] failed to create shard
	at org.elasticsearch.index.IndexService.createShard(IndexService.java:357)
	at org.elasticsearch.indices.cluster.IndicesClusterStateService.applyInitializingShard(IndicesClusterStateService.java:704)
	at org.elasticsearch.indices.cluster.IndicesClusterStateService.applyNewOrUpdatedShards(IndicesClusterStateService.java:605)
	at org.elasticsearch.indices.cluster.IndicesClusterStateService.clusterChanged(IndicesClusterStateService.java:185)
	at org.elasticsearch.cluster.service.InternalClusterService$UpdateTask.run(InternalClusterService.java:480)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:188)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:158)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.lucene.store.LockObtainFailedException: Can't lock shard [de_v4][2], timed out after 5000ms
	at org.elasticsearch.env.NodeEnvironment$InternalShardLock.acquire(NodeEnvironment.java:576)
	at org.elasticsearch.env.NodeEnvironment.shardLock(NodeEnvironment.java:504)
	at org.elasticsearch.index.IndexService.createShard(IndexService.java:310)
	... 9 more
[2015-06-22 14:48:27,309][WARN ][cluster.action.shard     ] [Grim Hunter] [de_v4][2] received shard failed for [de_v4][2], node[iTOuyuGSRLab6ZZxwnhb3g], [P], s[INITIALIZING], indexUUID [jsOkLa-_Q8GSPkKAXXfsoQ], reason [shard failure [failed to create shard][IndexShardCreationException[[de_v4][2] failed to create shard]; nested: LockObtainFailedException[Can't lock shard [de_v4][2], timed out after 5000ms]; ]]
[2015-06-22 14:48:32,309][WARN ][indices.cluster          ] [Grim Hunter] [[de_v4][0]] marking and sending shard failed due to [failed to create shard]
org.elasticsearch.index.shard.IndexShardCreationException: [de_v4][0] failed to create shard
	at org.elasticsearch.index.IndexService.createShard(IndexService.java:357)
	at org.elasticsearch.indices.cluster.IndicesClusterStateService.applyInitializingShard(IndicesClusterStateService.java:704)
	at org.elasticsearch.indices.cluster.IndicesClusterStateService.applyNewOrUpdatedShards(IndicesClusterStateService.java:605)
	at org.elasticsearch.indices.cluster.IndicesClusterStateService.clusterChanged(IndicesClusterStateService.java:185)
	at org.elasticsearch.cluster.service.InternalClusterService$UpdateTask.run(InternalClusterService.java:480)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:188)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:158)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.lucene.store.LockObtainFailedException: Can't lock shard [de_v4][0], timed out after 5000ms
	at org.elasticsearch.env.NodeEnvironment$InternalShardLock.acquire(NodeEnvironment.java:576)
	at org.elasticsearch.env.NodeEnvironment.shardLock(NodeEnvironment.java:504)
	at org.elasticsearch.index.IndexService.createShard(IndexService.java:310)
	... 9 more

[... trimmed ...]

[2015-06-22 14:49:17,414][WARN ][cluster.action.shard     ] [Grim Hunter] [de_v4][0] received shard failed for [de_v4][0], node[iTOuyuGSRLab6ZZxwnhb3g], [P], s[INITIALIZING], indexUUID [jsOkLa-_Q8GSPkKAXXfsoQ], reason [shard failure [failed to create shard][IndexShardCreationException[[de_v4][0] failed to create shard]; nested: LockObtainFailedException[Can't lock shard [de_v4][0], timed out after 5000ms]; ]]
[2015-06-22 14:49:17,414][WARN ][cluster.action.shard     ] [Grim Hunter] [de_v4][3] received shard failed for [de_v4][3], node[iTOuyuGSRLab6ZZxwnhb3g], [P], s[INITIALIZING], indexUUID [jsOkLa-_Q8GSPkKAXXfsoQ], reason [master [Grim Hunter][iTOuyuGSRLab6ZZxwnhb3g][u-excus][inet[/192.168.1.181:9300]] marked shard as initializing, but shard is marked as failed, resend shard failure]
```

*Update:*
Today I adjusted the logging settings and briefly before the cluster is getting green:
```
[2015-06-23 12:21:00,100][DEBUG][search                   ] [Prime Mover] freeing search context [1], time [1435054860015], lastAccessTime [1435054776777], keepAlive [60000]
[2015-06-23 12:21:00,101][DEBUG][search                   ] [Prime Mover] freeing search context [2], time [1435054860015], lastAccessTime [1435054776777], keepAlive [60000]
```

btw: [clearing](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-scroll.html#_clear_scroll_api) (manually) all scroll ids during closing / opening process solves the problem (as temporary workaround); cluster state will be green immediately after reopening the index.

```
curl -XDELETE localhost:9200/_search/scroll/_all
```

Tested with Elasticsearch v1.6.0

--- a/core/src/main/java/org/elasticsearch/search/SearchService.java
+++ b/core/src/main/java/org/elasticsearch/search/SearchService.java
@@ -159,7 +159,7 @@ public class SearchService extends AbstractLifecycleComponent<SearchService> {
         indicesService.indicesLifecycle().addListener(new IndicesLifecycle.Listener() {
 
             @Override
-            public void afterIndexDeleted(Index index, @IndexSettings Settings indexSettings) {
+            public void afterIndexClosed(Index index, @IndexSettings Settings indexSettings) {
                 // once an index is closed we can just clean up all the pending search context information
                 // to release memory and let references to the filesystem go etc.
                 freeAllContextForIndex(index);

--- a/core/src/test/java/org/elasticsearch/search/scroll/SearchScrollTests.java
+++ b/core/src/test/java/org/elasticsearch/search/scroll/SearchScrollTests.java
@@ -38,6 +38,7 @@ import org.elasticsearch.test.ElasticsearchIntegrationTest;
 import org.elasticsearch.test.hamcrest.ElasticsearchAssertions;
 import org.junit.Test;
 
+import java.io.IOException;
 import java.util.Map;
 
 import static org.elasticsearch.common.xcontent.XContentFactory.jsonBuilder;
@@ -573,6 +574,33 @@ public class SearchScrollTests extends ElasticsearchIntegrationTest {
         } catch (Exception e) {
             assertThat(e, instanceOf(IllegalArgumentException.class));
             assertThat(e.getMessage(), startsWith("Unknown parameter [unknown]"));
+        }
+    }
+
+    public void testCloseAndReopenOrDeleteWithActiveScroll() throws IOException {
+        createIndex("test");
+        for (int i = 0; i < 100; i++) {
+            client().prepareIndex("test", "type1", Integer.toString(i)).setSource(jsonBuilder().startObject().field("field", i).endObject()).execute().actionGet();
+        }
+        refresh();
+        SearchResponse searchResponse = client().prepareSearch()
+                .setQuery(matchAllQuery())
+                .setSize(35)
+                .setScroll(TimeValue.timeValueMinutes(2))
+                .addSort("field", SortOrder.ASC)
+                .execute().actionGet();
+        long counter = 0;
+        assertThat(searchResponse.getHits().getTotalHits(), equalTo(100l));
+        assertThat(searchResponse.getHits().hits().length, equalTo(35));
+        for (SearchHit hit : searchResponse.getHits()) {
+            assertThat(((Number) hit.sortValues()[0]).longValue(), equalTo(counter++));
+        }
+        if (randomBoolean()) {
+            client().admin().indices().prepareClose("test").get();
+            client().admin().indices().prepareOpen("test").get();
+            ensureGreen("test");
+        } else {
+            client().admin().indices().prepareDelete("test").get();
         }
     }
 

